From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: NONPLAYT <nonplay@bxteam.org>
Date: Thu, 10 Jul 2025 03:48:14 +0300
Subject: [PATCH] C2ME: Optimize Aquifer

This patch is based on the following mixins:
* "com/ishland/c2me/opts/worldgen/vanilla/mixin/aquifer/MixinAquiferSamplerImpl.java"
* "com/ishland/c2me/opts/worldgen/general/common/random_instances/RandomUtils.java"
By: ishland <ishlandmc@yeah.net>
As part of: C2ME (https://github.com/RelativityMC/C2ME-fabric)
Licensed under: MIT (https://opensource.org/licenses/MIT)

diff --git a/net/minecraft/world/level/levelgen/Aquifer.java b/net/minecraft/world/level/levelgen/Aquifer.java
index 7b53acc0c5f02bc1ae27852b36b6864baa1a72de..b23475c54211e91f5e8bdce5b841bc717ab810ee 100644
--- a/net/minecraft/world/level/levelgen/Aquifer.java
+++ b/net/minecraft/world/level/levelgen/Aquifer.java
@@ -123,6 +123,14 @@ public interface Aquifer {
         private final int gridSizeZ;
         private final DensitySampler.Bound surfaceLevel;
         private final Long2IntMap surfaceLevelCache = new Long2IntOpenHashMap();
+        // Leaf start - C2ME: Optimize Aquifer
+        private int c2me$packed1;
+        private int c2me$packed2;
+        private int c2me$packed3;
+        private int c2me$packed4;
+        private double c2me$mutableDoubleThingy;
+        private final short[] c2me$packedBlockPositions;
+        // Leaf end - C2ME: Optimize Aquifer
         private static final int[][] SURFACE_SAMPLING_OFFSETS_IN_CHUNKS = new int[][]{
             {0, 0}, {-2, -1}, {-1, -1}, {0, -1}, {1, -1}, {-3, 0}, {-2, 0}, {-1, 0}, {1, 0}, {-2, 1}, {-1, 1}, {0, 1}, {1, 1}
         };
@@ -155,6 +163,31 @@ public interface Aquifer {
             this.aquiferCache = new Aquifer.FluidStatus[totalGridSize];
             this.aquiferLocationCache = new long[totalGridSize];
             Arrays.fill(this.aquiferLocationCache, Long.MAX_VALUE);
+            // Leaf start - C2ME: Optimize Aquifer - preload aquifer locations
+            if (this.aquiferLocationCache.length % (this.gridSizeX * this.gridSizeZ) != 0) {
+                throw new AssertionError("Array length");
+            }
+
+            this.c2me$packedBlockPositions = new short[this.aquiferLocationCache.length];
+
+            final RandomSource random = com.ishland.c2me.opts.worldgen.general.common.random_instances.RandomUtils.getRandom(this.positionalRandomFactory);
+            for (int y = 0; y < gridSizeY; y++) {
+                for (int z = 0; z < this.gridSizeZ; z++) {
+                    for (int x = 0; x < this.gridSizeX; x++) {
+                        final int x1 = x + this.minGridX;
+                        final int y1 = y + this.minGridY;
+                        final int z1 = z + this.minGridZ;
+                        com.ishland.c2me.opts.worldgen.general.common.random_instances.RandomUtils.derive(this.positionalRandomFactory, random, x1, y1, z1);
+                        int r0 = random.nextInt(10);
+                        int r1 = random.nextInt(9);
+                        int r2 = random.nextInt(10);
+                        int index = this.getIndex(x1, y1, z1);
+                        this.aquiferLocationCache[index] = BlockPos.asLong(fromGridX(x1, r0), fromGridY(y1, r1), fromGridZ(z1, r2));
+                        this.c2me$packedBlockPositions[index] = (short) ((r0 << 8) | (r1 << 4) | r2);
+                    }
+                }
+            }
+            // Leaf end - C2ME: Optimize Aquifer - preload aquifer locations
             int maxAdjustedSurfaceLevel = this.adjustSurfaceLevel(
                 this.maxSurfaceLevel(fromGridX(this.minGridX, 0), fromGridZ(this.minGridZ, 0), fromGridX(maxGridX, 9), fromGridZ(maxGridZ, 9))
             );
@@ -220,135 +253,10 @@ public interface Aquifer {
                 return SharedConstants.DEBUG_DISABLE_FLUID_GENERATION ? Blocks.AIR.defaultBlockState() : Blocks.LAVA.defaultBlockState();
             }
 
-            int xAnchor = gridX(blockX + -5);
-            int yAnchor = gridY(blockY + 1);
-            int zAnchor = gridZ(blockZ + -5);
-            int distanceSqr1 = Integer.MAX_VALUE;
-            int distanceSqr2 = Integer.MAX_VALUE;
-            int distanceSqr3 = Integer.MAX_VALUE;
-            int distanceSqr4 = Integer.MAX_VALUE;
-            int closestIndex1 = 0;
-            int closestIndex2 = 0;
-            int closestIndex3 = 0;
-            int closestIndex4 = 0;
-
-            for (int x1 = 0; x1 <= 1; x1++) {
-                for (int y1 = -1; y1 <= 1; y1++) {
-                    for (int z1 = 0; z1 <= 1; z1++) {
-                        int spacedGridX = xAnchor + x1;
-                        int spacedGridY = yAnchor + y1;
-                        int spacedGridZ = zAnchor + z1;
-                        int index = this.getIndex(spacedGridX, spacedGridY, spacedGridZ);
-                        long existingLocation = this.aquiferLocationCache[index];
-                        long location;
-                        if (existingLocation != Long.MAX_VALUE) {
-                            location = existingLocation;
-                        } else {
-                            RandomSource random = this.positionalRandomFactory.at(spacedGridX, spacedGridY, spacedGridZ);
-                            location = BlockPos.asLong(
-                                fromGridX(spacedGridX, random.nextInt(10)),
-                                fromGridY(spacedGridY, random.nextInt(9)),
-                                fromGridZ(spacedGridZ, random.nextInt(10))
-                            );
-                            this.aquiferLocationCache[index] = location;
-                        }
-
-                        int dx = BlockPos.getX(location) - blockX;
-                        int dy = BlockPos.getY(location) - blockY;
-                        int dz = BlockPos.getZ(location) - blockZ;
-                        int newDistance = dx * dx + dy * dy + dz * dz;
-                        if (distanceSqr1 >= newDistance) {
-                            closestIndex4 = closestIndex3;
-                            closestIndex3 = closestIndex2;
-                            closestIndex2 = closestIndex1;
-                            closestIndex1 = index;
-                            distanceSqr4 = distanceSqr3;
-                            distanceSqr3 = distanceSqr2;
-                            distanceSqr2 = distanceSqr1;
-                            distanceSqr1 = newDistance;
-                        } else if (distanceSqr2 >= newDistance) {
-                            closestIndex4 = closestIndex3;
-                            closestIndex3 = closestIndex2;
-                            closestIndex2 = index;
-                            distanceSqr4 = distanceSqr3;
-                            distanceSqr3 = distanceSqr2;
-                            distanceSqr2 = newDistance;
-                        } else if (distanceSqr3 >= newDistance) {
-                            closestIndex4 = closestIndex3;
-                            closestIndex3 = index;
-                            distanceSqr4 = distanceSqr3;
-                            distanceSqr3 = newDistance;
-                        } else if (distanceSqr4 >= newDistance) {
-                            closestIndex4 = index;
-                            distanceSqr4 = newDistance;
-                        }
-                    }
-                }
-            }
-
-            Aquifer.FluidStatus closestStatus1 = this.getAquiferStatus(closestIndex1);
-            double similarity12 = similarity(distanceSqr1, distanceSqr2);
-            BlockState fluidState = closestStatus1.at(blockY);
-            BlockState actualFluidState = SharedConstants.DEBUG_DISABLE_FLUID_GENERATION ? Blocks.AIR.defaultBlockState() : fluidState;
-            if (similarity12 <= 0.0) {
-                if (similarity12 >= FLOWING_UPDATE_SIMULARITY) {
-                    Aquifer.FluidStatus closestStatus2 = this.getAquiferStatus(closestIndex2);
-                    this.shouldScheduleFluidUpdate = !closestStatus1.equals(closestStatus2);
-                } else {
-                    this.shouldScheduleFluidUpdate = false;
-                }
-
-                return actualFluidState;
-            } else {
-                if (fluidState.is(Blocks.WATER) && this.globalFluidPicker.computeFluid(blockX, blockY - 1, blockZ).at(blockY - 1).is(Blocks.LAVA)) {
-                    this.shouldScheduleFluidUpdate = true;
-                    return actualFluidState;
-                }
-
-                MutableDouble barrierNoiseValue = new MutableDouble(Double.NaN);
-                Aquifer.FluidStatus closestStatus2 = this.getAquiferStatus(closestIndex2);
-                double barrier12 = similarity12 * this.calculatePressure(blockX, blockY, blockZ, barrierNoiseValue, closestStatus1, closestStatus2);
-                if (density + barrier12 > 0.0) {
-                    this.shouldScheduleFluidUpdate = false;
-                    return null;
-                }
-
-                Aquifer.FluidStatus closestStatus3 = this.getAquiferStatus(closestIndex3);
-                double similarity13 = similarity(distanceSqr1, distanceSqr3);
-                if (similarity13 > 0.0) {
-                    double barrier13 = similarity12
-                        * similarity13
-                        * this.calculatePressure(blockX, blockY, blockZ, barrierNoiseValue, closestStatus1, closestStatus3);
-                    if (density + barrier13 > 0.0) {
-                        this.shouldScheduleFluidUpdate = false;
-                        return null;
-                    }
-                }
-
-                double similarity23 = similarity(distanceSqr2, distanceSqr3);
-                if (similarity23 > 0.0) {
-                    double barrier23 = similarity12
-                        * similarity23
-                        * this.calculatePressure(blockX, blockY, blockZ, barrierNoiseValue, closestStatus2, closestStatus3);
-                    if (density + barrier23 > 0.0) {
-                        this.shouldScheduleFluidUpdate = false;
-                        return null;
-                    }
-                }
-
-                boolean mayFlow12 = !closestStatus1.equals(closestStatus2);
-                boolean mayFlow23 = similarity23 >= FLOWING_UPDATE_SIMULARITY && !closestStatus2.equals(closestStatus3);
-                boolean mayFlow13 = similarity13 >= FLOWING_UPDATE_SIMULARITY && !closestStatus1.equals(closestStatus3);
-                if (!mayFlow12 && !mayFlow23 && !mayFlow13) {
-                    this.shouldScheduleFluidUpdate = similarity13 >= FLOWING_UPDATE_SIMULARITY
-                        && similarity(distanceSqr1, distanceSqr4) >= FLOWING_UPDATE_SIMULARITY
-                        && !closestStatus1.equals(this.getAquiferStatus(closestIndex4));
-                } else {
-                    this.shouldScheduleFluidUpdate = true;
-                }
-
-                return actualFluidState;
-            }
+            // Leaf start - C2ME: Optimize Aquifer
+            this.c2me$refreshClosest(blockX, blockY, blockZ);
+            return this.c2me$computeSubstancePost(blockX, blockY, blockZ, density);
+            // Leaf end - C2ME: Optimize Aquifer
         }
 
         @Override
@@ -425,6 +333,200 @@ public interface Aquifer {
             }
         }
 
+        // Leaf start - C2ME: Optimize Aquifer
+        private @Nullable BlockState c2me$computeSubstancePost(final int blockX, final int blockY, final int blockZ, final double density) {
+            Aquifer.FluidStatus closestStatus1 = this.getAquiferStatus(c2me$unpackPosIdx(this.c2me$packed1));
+            double similarity12 = similarity(c2me$unpackDist(this.c2me$packed1), c2me$unpackDist(this.c2me$packed2));
+            BlockState fluidState = closestStatus1.at(blockY);
+            BlockState actualFluidState = SharedConstants.DEBUG_DISABLE_FLUID_GENERATION ? Blocks.AIR.defaultBlockState() : fluidState;
+            if (similarity12 <= 0.0) {
+                if (similarity12 >= FLOWING_UPDATE_SIMULARITY) {
+                    Aquifer.FluidStatus closestStatus2 = this.getAquiferStatus(c2me$unpackPosIdx(this.c2me$packed2));
+                    this.shouldScheduleFluidUpdate = !closestStatus1.equals(closestStatus2);
+                } else {
+                    this.shouldScheduleFluidUpdate = false;
+                }
+
+                return actualFluidState;
+            } else if (fluidState.is(Blocks.WATER) && this.globalFluidPicker.computeFluid(blockX, blockY - 1, blockZ).at(blockY - 1).is(Blocks.LAVA)) {
+                this.shouldScheduleFluidUpdate = true;
+                return actualFluidState;
+            }
+
+            this.c2me$mutableDoubleThingy = Double.NaN;
+            Aquifer.FluidStatus closestStatus2 = this.getAquiferStatus(c2me$unpackPosIdx(this.c2me$packed2));
+            double barrier12 = similarity12 * this.c2me$calculatePressure(blockX, blockY, blockZ, closestStatus1, closestStatus2);
+            if (density + barrier12 > 0.0) {
+                this.shouldScheduleFluidUpdate = false;
+                return null;
+            }
+
+            Aquifer.FluidStatus closestStatus3 = this.getAquiferStatus(c2me$unpackPosIdx(this.c2me$packed3));
+            int dist1 = c2me$unpackDist(this.c2me$packed1);
+            int dist2 = c2me$unpackDist(this.c2me$packed2);
+            int dist3 = c2me$unpackDist(this.c2me$packed3);
+            double similarity13 = similarity(dist1, dist3);
+            if (this.c2me$checkBarrier(blockX, blockY, blockZ, density, similarity12, similarity13, closestStatus1, closestStatus3)) {
+                return null;
+            }
+
+            double similarity23 = similarity(dist2, dist3);
+            if (this.c2me$checkBarrier(blockX, blockY, blockZ, density, similarity12, similarity23, closestStatus2, closestStatus3)) {
+                return null;
+            }
+
+            boolean mayFlow12 = !closestStatus1.equals(closestStatus2);
+            boolean mayFlow23 = similarity23 >= FLOWING_UPDATE_SIMULARITY && !closestStatus2.equals(closestStatus3);
+            boolean mayFlow13 = similarity13 >= FLOWING_UPDATE_SIMULARITY && !closestStatus1.equals(closestStatus3);
+            if (!mayFlow12 && !mayFlow23 && !mayFlow13) {
+                this.shouldScheduleFluidUpdate = similarity13 >= FLOWING_UPDATE_SIMULARITY
+                    && similarity(dist1, c2me$unpackDist(this.c2me$packed4)) >= FLOWING_UPDATE_SIMULARITY
+                    && !closestStatus1.equals(this.getAquiferStatus(c2me$unpackPosIdx(this.c2me$packed4)));
+            } else {
+                this.shouldScheduleFluidUpdate = true;
+            }
+
+            return actualFluidState;
+        }
+
+        private boolean c2me$checkBarrier(
+            final int blockX, final int blockY, final int blockZ, final double density,
+            final double similarity12, final double similarity, final Aquifer.FluidStatus statusA, final Aquifer.FluidStatus statusB
+        ) {
+            if (similarity > 0.0) {
+                double barrier = similarity12 * similarity * this.c2me$calculatePressure(blockX, blockY, blockZ, statusA, statusB);
+                if (density + barrier > 0.0) {
+                    this.shouldScheduleFluidUpdate = false;
+                    return true;
+                }
+            }
+            return false;
+        }
+
+        // Branchless 4-nearest search over the 12 candidate aquifer centers. Each candidate is packed as
+        // (distSqr << 20) | (tieBreak << 16) | posIdx so a single int min/max network keeps them sorted.
+        // tieBreak = 11 - (vanilla visit order), so on equal distance the later-visited sample wins, exactly like vanilla's ">=".
+        private void c2me$refreshClosest(final int blockX, final int blockY, final int blockZ) {
+            int gridX = gridX(blockX + -5);
+            int gridY = gridY(blockY + 1);
+            int gridZ = gridZ(blockZ + -5);
+            int closest1 = Integer.MAX_VALUE;
+            int closest2 = Integer.MAX_VALUE;
+            int closest3 = Integer.MAX_VALUE;
+            int closest4 = Integer.MAX_VALUE;
+
+            for (int offY = -1; offY <= 1; ++offY) {
+                int gridYBlockBase = (gridY + offY) * 12;
+                for (int offZ = 0; offZ <= 1; ++offZ) {
+                    int gridZBlockBase = (gridZ + offZ) << 4;
+                    int visitOrderYZ = (offY + 1) * 2 + offZ;
+
+                    int leftPosIdx = this.getIndex(gridX, gridY + offY, gridZ + offZ);
+                    int leftPacked = this.c2me$packedBlockPositions[leftPosIdx];
+                    int leftDx = (gridX << 4) + c2me$unpackX(leftPacked) - blockX;
+                    int leftDy = gridYBlockBase + c2me$unpackY(leftPacked) - blockY;
+                    int leftDz = gridZBlockBase + c2me$unpackZ(leftPacked) - blockZ;
+                    int leftKey = ((leftDx * leftDx + leftDy * leftDy + leftDz * leftDz) << 20) | ((11 - visitOrderYZ) << 16) | leftPosIdx;
+
+                    int rightPosIdx = leftPosIdx + 1;
+                    int rightPacked = this.c2me$packedBlockPositions[rightPosIdx];
+                    int rightDx = ((gridX + 1) << 4) + c2me$unpackX(rightPacked) - blockX;
+                    int rightDy = gridYBlockBase + c2me$unpackY(rightPacked) - blockY;
+                    int rightDz = gridZBlockBase + c2me$unpackZ(rightPacked) - blockZ;
+                    int rightKey = ((rightDx * rightDx + rightDy * rightDy + rightDz * rightDz) << 20) | ((5 - visitOrderYZ) << 16) | rightPosIdx;
+
+                    if (leftKey <= closest4) {
+                        int n1 = Math.max(closest1, leftKey);
+                        closest1 = Math.min(closest1, leftKey);
+                        int n2 = Math.max(closest2, n1);
+                        closest2 = Math.min(closest2, n1);
+                        int n3 = Math.max(closest3, n2);
+                        closest3 = Math.min(closest3, n2);
+                        closest4 = Math.min(closest4, n3);
+                    }
+
+                    if (rightKey <= closest4) {
+                        int n1 = Math.max(closest1, rightKey);
+                        closest1 = Math.min(closest1, rightKey);
+                        int n2 = Math.max(closest2, n1);
+                        closest2 = Math.min(closest2, n1);
+                        int n3 = Math.max(closest3, n2);
+                        closest3 = Math.min(closest3, n2);
+                        closest4 = Math.min(closest4, n3);
+                    }
+                }
+            }
+
+            this.c2me$packed1 = closest1;
+            this.c2me$packed2 = closest2;
+            this.c2me$packed3 = closest3;
+            this.c2me$packed4 = closest4;
+        }
+
+        private double c2me$calculatePressure(
+            final int blockX, final int blockY, final int blockZ, final Aquifer.FluidStatus statusClosest1, final Aquifer.FluidStatus statusClosest2
+        ) {
+            BlockState type1 = statusClosest1.at(blockY);
+            BlockState type2 = statusClosest2.at(blockY);
+            if ((!type1.is(Blocks.LAVA) || !type2.is(Blocks.WATER)) && (!type1.is(Blocks.WATER) || !type2.is(Blocks.LAVA))) {
+                int fluidYDiff = Math.abs(statusClosest1.fluidLevel - statusClosest2.fluidLevel);
+                if (fluidYDiff == 0) {
+                    return 0.0;
+                }
+
+                double averageFluidY = 0.5 * (statusClosest1.fluidLevel + statusClosest2.fluidLevel);
+                double gradient = c2me$pressureGradient(blockY, averageFluidY, fluidYDiff);
+                double noiseValue;
+                if (!(gradient < -2.0) && !(gradient > 2.0)) {
+                    double cached = this.c2me$mutableDoubleThingy;
+                    if (Double.isNaN(cached)) {
+                        noiseValue = this.barrierNoise.sampleValue(blockX, blockY, blockZ);
+                        this.c2me$mutableDoubleThingy = noiseValue;
+                    } else {
+                        noiseValue = cached;
+                    }
+                } else {
+                    noiseValue = 0.0;
+                }
+
+                return 2.0 * (noiseValue + gradient);
+            } else {
+                return 2.0;
+            }
+        }
+
+        private static double c2me$pressureGradient(final int blockY, final double averageFluidY, final int fluidYDiff) {
+            double howFarAboveAverageFluidPoint = blockY + 0.5 - averageFluidY;
+            double distanceFromBarrierEdgeTowardsMiddle = fluidYDiff / 2.0 - Math.abs(howFarAboveAverageFluidPoint);
+            if (howFarAboveAverageFluidPoint > 0.0) {
+                return distanceFromBarrierEdgeTowardsMiddle > 0.0 ? distanceFromBarrierEdgeTowardsMiddle / 1.5 : distanceFromBarrierEdgeTowardsMiddle / 2.5;
+            } else {
+                double centerPoint = 3.0 + distanceFromBarrierEdgeTowardsMiddle;
+                return centerPoint > 0.0 ? centerPoint / 3.0 : centerPoint / 10.0;
+            }
+        }
+
+        private static int c2me$unpackX(final int packed) {
+            return packed >> 8;
+        }
+
+        private static int c2me$unpackY(final int packed) {
+            return (packed >> 4) & 0b1111;
+        }
+
+        private static int c2me$unpackZ(final int packed) {
+            return packed & 0b1111;
+        }
+
+        private static int c2me$unpackDist(final int packed) {
+            return packed >> 20;
+        }
+
+        private static int c2me$unpackPosIdx(final int packed) {
+            return packed & 0xffff;
+        }
+        // Leaf end - C2ME: Optimize Aquifer
+
         private static int gridX(final int blockCoord) {
             return blockCoord >> 4;
         }
@@ -518,13 +620,12 @@ public interface Aquifer {
                 fullyFloodidness = -1.0;
             } else {
                 int distanceBelowSurface = this.adjustSurfaceLevel(lowestSurfaceLevel) - y;
-                int floodednessMaxDepth = 64;
-                double floodednessFactor = surfaceAtCenterIsUnderGlobalFluidLevel ? Mth.clampedMap(distanceBelowSurface, 0.0, 64.0, 1.0, 0.0) : 0.0;
+                // Leaf start - C2ME: Optimize Aquifer - inline Mth.clampedMap and Mth.map
+                double floodednessFactor = surfaceAtCenterIsUnderGlobalFluidLevel ? Mth.clampedLerp(distanceBelowSurface / 64.0, 1.0, 0.0) : 0.0;
                 double floodednessNoiseValue = Mth.clamp(this.fluidLevelFloodednessNoise.sampleValue(x, y, z), -1.0, 1.0);
-                double fullyFloodedThreshold = Mth.map(floodednessFactor, 1.0, 0.0, -0.3, 0.8);
-                double partiallyFloodedThreshold = Mth.map(floodednessFactor, 1.0, 0.0, -0.8, 0.4);
-                partiallyFloodedness = floodednessNoiseValue - partiallyFloodedThreshold;
-                fullyFloodidness = floodednessNoiseValue - fullyFloodedThreshold;
+                partiallyFloodedness = floodednessNoiseValue + 0.8 + (floodednessFactor - 1.0) * 1.2;
+                fullyFloodidness = floodednessNoiseValue + 0.3 + (floodednessFactor - 1.0) * 1.1;
+                // Leaf end - C2ME: Optimize Aquifer - inline Mth.clampedMap and Mth.map
             }
 
             int fluidSurfaceLevel;
@@ -556,11 +657,11 @@ public interface Aquifer {
         private BlockState computeFluidType(final int x, final int y, final int z, final Aquifer.FluidStatus globalFluid, final int fluidSurfaceLevel) {
             BlockState fluidType = globalFluid.fluidType;
             if (fluidSurfaceLevel <= -10 && fluidSurfaceLevel != DimensionType.WAY_BELOW_MIN_Y && globalFluid.fluidType != Blocks.LAVA.defaultBlockState()) {
-                int fluidTypeCellWidth = 64;
-                int fluidTypeCellHeight = 40;
-                int fluidTypeCellX = Math.floorDiv(x, 64);
+                // Leaf start - C2ME: Optimize Aquifer - use bit shift instead of Math.floorDiv for power-of-two divisors
+                int fluidTypeCellX = x >> 6;
                 int fluidTypeCellY = Math.floorDiv(y, 40);
-                int fluidTypeCellZ = Math.floorDiv(z, 64);
+                int fluidTypeCellZ = z >> 6;
+                // Leaf end - C2ME: Optimize Aquifer - use bit shift instead of Math.floorDiv for power-of-two divisors
                 double lavaNoiseValue = this.lavaNoise.sampleValue(fluidTypeCellX, fluidTypeCellY, fluidTypeCellZ);
                 if (Math.abs(lavaNoiseValue) > 0.3) {
                     fluidType = Blocks.LAVA.defaultBlockState();
diff --git a/net/minecraft/world/level/levelgen/LegacyRandomSource.java b/net/minecraft/world/level/levelgen/LegacyRandomSource.java
index 81e02ed613d8122074be7893d7ba33241e6137b2..b66bc1f27dfb6b5cf178a6bd7c8d04d5d3fbd88a 100644
--- a/net/minecraft/world/level/levelgen/LegacyRandomSource.java
+++ b/net/minecraft/world/level/levelgen/LegacyRandomSource.java
@@ -53,13 +53,7 @@ public class LegacyRandomSource implements BitRandomSource {
         return this.gaussianSource.nextGaussian();
     }
 
-    public static class LegacyPositionalRandomFactory implements PositionalRandomFactory {
-        private final long seed;
-
-        public LegacyPositionalRandomFactory(final long seed) {
-            this.seed = seed;
-        }
-
+    public record LegacyPositionalRandomFactory(long seed) implements PositionalRandomFactory { // Leaf - make record
         @Override
         public RandomSource at(final int x, final int y, final int z) {
             long positionalSeed = Mth.getSeed(x, y, z);
diff --git a/net/minecraft/world/level/levelgen/Xoroshiro128PlusPlus.java b/net/minecraft/world/level/levelgen/Xoroshiro128PlusPlus.java
index c002338d1efcf0e07b544e676ed3fc2ba50cd493..197c56785cba47719da770461f2e58870b2daada 100644
--- a/net/minecraft/world/level/levelgen/Xoroshiro128PlusPlus.java
+++ b/net/minecraft/world/level/levelgen/Xoroshiro128PlusPlus.java
@@ -5,8 +5,8 @@ import java.util.stream.LongStream;
 import net.minecraft.util.Util;
 
 public class Xoroshiro128PlusPlus {
-    private long seedLo;
-    private long seedHi;
+    public long seedLo;
+    public long seedHi;
     public static final Codec<Xoroshiro128PlusPlus> CODEC = Codec.LONG_STREAM
         .comapFlatMap(seed -> Util.fixedSize(seed, 2).map(longs -> new Xoroshiro128PlusPlus(longs[0], longs[1])), r -> LongStream.of(r.seedLo, r.seedHi));
 
diff --git a/net/minecraft/world/level/levelgen/XoroshiroRandomSource.java b/net/minecraft/world/level/levelgen/XoroshiroRandomSource.java
index fe3f1777546504cab9c9d1fcf7d5c2efe59d100a..eeabdcb3ebf9382fc75252acbc64d0ceadb11843 100644
--- a/net/minecraft/world/level/levelgen/XoroshiroRandomSource.java
+++ b/net/minecraft/world/level/levelgen/XoroshiroRandomSource.java
@@ -10,7 +10,7 @@ public class XoroshiroRandomSource implements RandomSource {
     private static final double DOUBLE_UNIT = 1.110223E-16F;
     public static final Codec<XoroshiroRandomSource> CODEC = Xoroshiro128PlusPlus.CODEC
         .xmap(generator -> new XoroshiroRandomSource(generator), source -> source.randomNumberGenerator);
-    private Xoroshiro128PlusPlus randomNumberGenerator;
+    public Xoroshiro128PlusPlus randomNumberGenerator;
     private final MarsagliaPolarGaussian gaussianSource = new MarsagliaPolarGaussian(this);
 
     public XoroshiroRandomSource(final long seed) {
@@ -109,15 +109,7 @@ public class XoroshiroRandomSource implements RandomSource {
         return this.randomNumberGenerator.nextLong() >>> 64 - bits;
     }
 
-    public static class XoroshiroPositionalRandomFactory implements PositionalRandomFactory {
-        private final long seedLo;
-        private final long seedHi;
-
-        public XoroshiroPositionalRandomFactory(final long seedLo, final long seedHi) {
-            this.seedLo = seedLo;
-            this.seedHi = seedHi;
-        }
-
+    public record XoroshiroPositionalRandomFactory(long seedLo, long seedHi) implements PositionalRandomFactory { // Leaf - make record
         @Override
         public RandomSource at(final int x, final int y, final int z) {
             long positionalSeed = Mth.getSeed(x, y, z);
